src/xconnection.cpp

```diff
@@ -389,22 +385,63 @@ std::experimental::optional<vector<Window>>
 std::experimental::optional<vector<string>>
     XConnection::getWindowPropertyTextList(Window window, Atom property)
 {
-    XTextProperty text_prop;
-    if (!XGetTextProperty(m_display, window, &text_prop, property)) {
+    Atom prop_type;
+    int format;
+    unsigned long bytes_left;
+    char* items_return; // TODO: char* or unsigned char*?
+    unsigned long count;
+    int status = XGetWindowProperty(m_display, window,
+            property, 0, ULONG_MAX, False, AnyPropertyType,
+            &prop_type, &format, &count, &bytes_left,
+            (unsigned char**)&items_return);
+    if (Success != status || prop_type == None || format == 0) {
         return {};
     }
-    char** list_return;
-    int count;
-    if (Success != Xutf8TextPropertyToTextList(m_display, &text_prop, &list_return, &count)) {
-        XFree(text_prop.value);
-        return {};
+    if (format != 8) {
+        fprintf(stderr, "herbstluftwm error: can not parse the "
+                        " atom \'%s\' of window 0x%lx: expected format=8 but got"
+                        " format=%d\n",
+                        XGetAtomName(m_display, property), window, format);
+        XFree(items_return);
     }
+    unsigned long offset = 0;
     vector<string> arguments;
-    for (int i = 0; i < count; i++) {
-        arguments.push_back(list_return[i]);
+    // the trailing 0 at items_return[count] might be crucial:
+    // if the string list ends with the empty string, then we
+    // need to access items_return[count].
+    while (offset <= count) {
+        char* textChunk = items_return + offset;
+        // let us hope that items_return is properly null-byte terminated.
+        unsigned long textChunkLen = strlen(textChunk);
+        // copy into a new string object and convert to utf8 if necessary:
+        if (prop_type == XA_STRING) {
+            // a XA_STRING is always encoded in ISO 8859-1
+            arguments.push_back(iso_8859_1_to_utf8(textChunk));
+        } else if (prop_type == utf8StringAtom_) {
+            arguments.push_back(textChunk);
+        } else {
+            // try to convert via XmbTextPropertyToTextList, just like
+            // xprop does:
+            XTextProperty textprop;
+            textprop.encoding = prop_type;
+            textprop.format = 8;
+            textprop.nitems = textChunkLen + 1;
+            textprop.value = reinterpret_cast<unsigned char*>(textChunk);
+            int n = 0;
+            char** list = nullptr;
+            if (XmbTextPropertyToTextList(m_display, &textprop, &list, &n) == Success
+                && n > 0 && *list)
+            {
+                arguments.push_back(*list);
+                XFreeStringList(list);
+            } else {
+                fprintf(stderr, "herbstluftwm warning: can not read text property\n");
```

Comment:
Is the mix of fprintf and std::cerr and HSDebug in this file intended?
Also "herbstluftwm warning" deviates from the rest of fprintf's patterns,
usually you see "herbstluftm: warning" or something?

Reply:
With these long format strings, I find fprintf more readable, what do you think?

Reply:
I'm fine with either, just wanted to check if the "inconsistency" was on purpse.